robot-server/robot_server/runs/run_store.py

```
@@ -80,7 +84,11 @@ def insert(self, run: RunResource) -> RunResource:
             _convert_run_to_sql_values(run=run)
         )
         with self._sql_engine.begin() as transaction:
-            transaction.execute(statement)
+            try:
+                transaction.execute(statement)
+            except sqlalchemy.exc.IntegrityError:
+                assert run.protocol_id is not None
```

This strikes me as a little iffy; is there any way we can pull out the specific integrity error rather than raise a 500 due to an AssertionError

---

We can remove the FK constraint and add the logic ourselves like the has implementation you add in the analysis PR. One more way is to convert the assert to an if and raise ProtocolNotFoundError else raise a different exception

---

> Still need to add a catch to the `runs/routers/base_router.py` for the `ProtocolNotFoundError` on run insert. Currently the endpoint will 500. This code path is quite difficult to hit regularly, but could conceivably be hit by concurrent `DELETE /protocols/:id` and `POST /runs` calls

We don't want to propagate this exception to the client? what is expected to happen in the catch block?

---

Discussed in person!

- Catch `ProtocolNotFoundError` from `run_store.insert` in the router to respond with a 404
- Add an assertion message to this assert to make the code a little easier to read and follow, but otherwise the 500 is actually what we want here